src/dotnet/OpenAPI.Validator/Validation/ArmResourcePropertiesBag.cs

```diff
@@ -47,19 +48,64 @@ public class ArmResourcePropertiesBag : TypedRule<Dictionary<string, Schema>>
         {
-            var resModels = context.ResourceModels;
-            var violatingModels = resModels.Where(res => definitions[res].Properties?.ContainsKey("properties") == true
-                                                         && definitions[res].Properties["properties"]?.Properties?.Keys.Intersect(ArmPropertiesBag).Any() == true);
-            foreach (var violatingModel in violatingModels)
+            IEnumerable<string> resourceModelsWithPropertiesBag = context.ResourceModels.Where(resourceModel => definitions[resourceModel].Properties?.ContainsKey("properties") == true);
+            foreach(string resourceModelName in resourceModelsWithPropertiesBag)
             {
-                var violatingProperties = definitions[violatingModel].Properties.Keys.Union(definitions[violatingModel].Properties["properties"].Properties.Keys).Intersect(ArmPropertiesBag);
-                yield return new ValidationMessage(new FileObjectPath(context.File, context.Path.AppendProperty(violatingModel).AppendProperty("properties")), this, violatingModel,
-                                                   string.Join(", ", violatingProperties));
+                Dictionary<string, IEnumerable<string>> violations = new Dictionary<string, IEnumerable<string>>();             
+                CheckModelForViolation(definitions[resourceModelName], resourceModelName, definitions, context, violations);
+
+                foreach (string modelName in violations.Keys)
+                {
+                    yield return new ValidationMessage(new FileObjectPath(context.File,
+                    context.Path.AppendProperty(modelName).AppendProperty("properties")), this, modelName,
+                                                   string.Join(", ", violations[modelName]));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Method to check the model for violation of this rule.
+        /// This method checks if the model has propeties bag. If so and it
+        /// has reference then the reference is processed. Else, the individual
+        /// properties are checked for violation.
+        ///
+        /// Then, if the model has allOf's, they need to be processed individually.
+        /// </summary>
+        /// <param name="resourceModel">schema of the model</param>
+        /// <param name="resourceModelName">name of the model</param>
+        /// <param name="definitions">definitions</param>
+        /// <param name="context">context</param>
+        /// <param name="violations">violations table</param>
+        private void CheckModelForViolation(Schema resourceModel, string resourceModelName, Dictionary<string, Schema> definitions, RuleContext context, Dictionary<string, IEnumerable<string>> violations)
+        {
+            if (resourceModel.Properties?.ContainsKey("properties") == true)
```

Comment:
Added the null check. But it is not required. Because, the null will be returned if the schema is referenced and not available. In that case code model will not be generated at all (errored out) and will not come to this rule. To be on the safe side, added the null check